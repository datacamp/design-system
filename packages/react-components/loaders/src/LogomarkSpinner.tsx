import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { ClassNames, keyframes } from '@emotion/react';
import React from 'react';

const stroke = keyframes({ '100%': { strokeDashoffset: 0 } });

const logoStartMask = keyframes({
  '5%': {
    clipPath: 'polygon(0% -12%, 0% -12%, 169% 63%, 169% 63%)',
  },
  '100%': {
    clipPath: 'polygon(0% -12%, 0% 308%, 169% 383%, 169% 63%)',
  },
});

const logoEndMask = keyframes({
  '71%': {
    clipPath: 'polygon(0% 0%, 0% 77%, 100% 34%, 100% -44%)',
  },
  '95%': {
    clipPath: 'polygon(0% 0%, 0% 140%, 100% 96%, 100% -44%)',
  },
});

const animationSettings = '2s infinite alternate';

interface LogomarkSpinner {
  className?: string;
  color?: 'navy' | 'white';
  size?: string;
}

const LogomarkSpinner = ({
  className,
  color = 'navy',
  size = '100px',
}: LogomarkSpinner): React.ReactElement => (
  // In Safari clip-path on SVG elements seems to break
  <div css={{ position: 'relative', width: size }}>
    <div css={{ paddingTop: '100%' }}>
      <ClassNames>
        {({ css, cx }) => (
          <div
            className={cx(
              className,
              css({
                animation: `${logoStartMask} cubic-bezier(0.65, 0, 0.55, 1) ${animationSettings}`,
                bottom: 0,
                left: '11.675%',
                position: 'absolute',
                right: '11.675%',
                top: 0,
              }),
            )}
          >
            <div
              css={{
                animation: `${logoEndMask} cubic-bezier(0.3, 0.75, 1, 0.85) ${animationSettings}`,
              }}
            >
              <div
                css={{
                  clipPath: 'polygon(0% -10%, 168.5% 64%, 0% 135%)',
                }}
              >
                <svg
                  version="1.1"
                  viewBox="0 0 2640 3444"
                  xmlns="http://www.w3.org/2000/svg"
                  xmlnsXlink="http://www.w3.org/1999/xlink"
                >
                  <title>Loading</title>
                  <path
                    css={{
                      animation: `${stroke} cubic-bezier(0.65, 0, 0.55, 1) ${animationSettings}`,
                      strokeDasharray: 9800,
                      strokeDashoffset: 9800,
                    }}
                    d="M0 0 M2569 1056L143 2447V149l1175 673v1867l1248 715"
                    fill="none"
                    stroke={tokens.color.primary[color].value.hex}
                    strokeLinejoin="round"
                    strokeWidth="300"
                  />
                </svg>
              </div>
            </div>
          </div>
        )}
      </ClassNames>
    </div>
  </div>
);

export default LogomarkSpinner;
