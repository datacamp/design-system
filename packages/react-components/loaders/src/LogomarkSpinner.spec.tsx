import { render } from '@testing-library/react';

import LogomarkSpinner from './LogomarkSpinner';

describe('LogomarkSpinner', () => {
  it('renders with default props', () => {
    const { container } = render(<LogomarkSpinner />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a white spinner', () => {
    const { container } = render(<LogomarkSpinner color="white" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders with a custom size', () => {
    const { container } = render(<LogomarkSpinner size="200px" />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
