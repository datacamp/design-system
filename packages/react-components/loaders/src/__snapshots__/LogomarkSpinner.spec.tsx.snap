// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogomarkSpinner renders with a custom color 1`] = `
@keyframes animation-0 {
  0% {
    stroke-dashoffset: 340%;
  }

  100% {
    stroke-dashoffset: 0;
  }
}

.emotion-0 {
  -webkit-animation: animation-0 2s infinite cubic-bezier(0.65,0,0.35,1) reverse;
  animation: animation-0 2s infinite cubic-bezier(0.65,0,0.35,1) reverse;
  stroke-dasharray: 170%;
  stroke-dashoffset: 0;
}

<svg
  height="100px"
  version="1.1"
  viewBox="0 0 225 225"
  width="100px"
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <title>
    Loading
  </title>
  <g
    fill="none"
    fill-rule="evenodd"
    stroke="none"
    stroke-width="1"
  >
    <polyline
      class="emotion-0"
      points="158.573452 177.08 112.038068 150.488473 112.038068 73.7151963 66 48 66 138.609263 157.803171 87.5483322"
      stroke="red"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="12"
    />
  </g>
</svg>
`;

exports[`LogomarkSpinner renders with a custom size 1`] = `
@keyframes animation-0 {
  0% {
    stroke-dashoffset: 340%;
  }

  100% {
    stroke-dashoffset: 0;
  }
}

.emotion-0 {
  -webkit-animation: animation-0 2s infinite cubic-bezier(0.65,0,0.35,1) reverse;
  animation: animation-0 2s infinite cubic-bezier(0.65,0,0.35,1) reverse;
  stroke-dasharray: 170%;
  stroke-dashoffset: 0;
}

<svg
  height="200px"
  version="1.1"
  viewBox="0 0 225 225"
  width="200px"
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <title>
    Loading
  </title>
  <g
    fill="none"
    fill-rule="evenodd"
    stroke="none"
    stroke-width="1"
  >
    <polyline
      class="emotion-0"
      points="158.573452 177.08 112.038068 150.488473 112.038068 73.7151963 66 48 66 138.609263 157.803171 87.5483322"
      stroke="#05192d"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="12"
    />
  </g>
</svg>
`;

exports[`LogomarkSpinner renders with default props 1`] = `
@keyframes animation-0 {
  0% {
    stroke-dashoffset: 340%;
  }

  100% {
    stroke-dashoffset: 0;
  }
}

.emotion-0 {
  -webkit-animation: animation-0 2s infinite cubic-bezier(0.65,0,0.35,1) reverse;
  animation: animation-0 2s infinite cubic-bezier(0.65,0,0.35,1) reverse;
  stroke-dasharray: 170%;
  stroke-dashoffset: 0;
}

<svg
  height="100px"
  version="1.1"
  viewBox="0 0 225 225"
  width="100px"
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <title>
    Loading
  </title>
  <g
    fill="none"
    fill-rule="evenodd"
    stroke="none"
    stroke-width="1"
  >
    <polyline
      class="emotion-0"
      points="158.573452 177.08 112.038068 150.488473 112.038068 73.7151963 66 48 66 138.609263 157.803171 87.5483322"
      stroke="#05192d"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="12"
    />
  </g>
</svg>
`;
